### DISCUSSION TITLE
Customize NLOHMANN_DEFINE_TYPE_INTRUSIVE

### DISCUSSION THREAD

**Discussion by lanyizi** [nlohmann/json#3350] [Q&A] [answered]
Hi! Macros like `NLOHMANN_DEFINE_TYPE_INTRUSIVE` really save a lot of typing for me.
But is there any way to "customize" it?
I have a class that needs to use mutex to prevent data race.
However, I can't add the lock statements inside the functions defined by `NLOHMANN_DEFINE_TYPE_INTRUSIVE`.

Then I thought I could provide a thread safe version of `Class::serialize()` / `Class::deserialize()` functions (which converts to / from json internally, while being locked with mutex), and have users use those function instead.
But I soon realized that `NLOHMANN_DEFINE_TYPE_INTRUSIVE` defines `friend` functions, which cannot be `private` since they are not member functions, so I can do nothing to prevent people from serializing / deserializing my class "directly".
What should I do in this case?

**Comment by nlohmann** [comment, marked as answer]
I guess best is to define the `to_json`/`from_json` functions yourself, see https://json.nlohmann.me/features/arbitrary_types/.

**Reply by lanyizi** [reply]
Looks like I must type more characters :)
Thanks for the answer!
